add option to configure s3 mappings provider

clemensvonschwerin opened the pull request:
Hey guys,

this PR addresses issue #6187. The provider for the S3 security mappings is made customisable.

clemensvonschwerin commented:
I signed the CLA today before creating the PR and sent an email, please check for it.

kokosing commented on presto-hive/src/main/java/io/prestosql/plugin/hive/s3/S3SecurityMappingConfig.java:
```diff
@@ -27,6 +27,20 @@
     private String roleCredentialName;
     private Duration refreshPeriod;
     private String colonReplacement;
+    private String securityMappingsProvider;
```
Use `enum` instead of `String` with values like `FILE`.

clemensvonschwerin replied:
This is relevant under the conditions that all implementations of `S3SecurityMappingsProvider` are part of `presto-hive`, right?

kokosing commented on presto-hive/src/main/java/io/prestosql/plugin/hive/s3/S3SecurityMappingConfig.java:
```diff
@@ -27,6 +27,20 @@
+    }
+
+    @Config("hive.s3.security-mapping.provider")
+    @ConfigDescription("class used to obtain the s3 security mappings")
```
We do not use reflection for things like that. We need to have all provider implementations here.

clemensvonschwerin replied:
So what I did locally to test the implementation and to setup for later is: I set up a new maven project, added `presto-hive` as a dependency and implemented a custom `S3SecurityMappingsProvider` in that project. Then I built the jar and put it into the hive-hadoop2 plugin directory. That way it can be found on the classpath and the custom implementation can be used. I think it leaves you more flexible in terms of implementation.

clemensvonschwerin commented:
Any thoughts on my comments?

alubbe commented:
Can someone re-run CI because `verification/cla-signed` should be green now? Or @clemensvonschwerin could you add a commit or rebase?

ebyhr commented:
@cla-bot check

clemensvonschwerin commented:
Hey guys, it has been a while. Any chance we can make some progress here? I would of course resolve the conflicts if there is a chance we can get this merged.

AdrianSkierniewski commented:
It would be really nice if we could merge this one :)

electrum commented:
Sorry for the delay, I missed this one originally. Can you explain more about your use case? Needing to drop in a JAR that depends on internal implementation details should be a last resort.

I can see a couple of ways this can be done without needing custom code:

1. Write the mapping to a local file. The file can be refreshed by specifying `hive.s3.security-mapping.refresh-period`, so you could an external process that periodically pulls the config from your service and writes it to a file.
2. Add a `hive.s3.security-mapping.config-uri` that would fetch from an HTTP location rather than a local file. This URL would return the same JSON format. This would allow you to run a simple proxy that fetches the config from your custom service and returns it to Trino.

Either of these approaches avoids the need to write and maintain custom code that can break between Trino releases. You can integrate with your config service in whatever language and way that is easiest.

I'm `@david` on the [Trino Slack](https://trino.io/slack.html) if it's easier to discuss there.

clemensvonschwerin commented:
@cla-bot check

clemensvonschwerin commented:
Hi guys, it would be awesome if we could get this one closed.

clemensvonschwerin commented:
> Sorry for the delay, I missed this one originally. Can you explain more about your use case? Needing to drop in a JAR that depends on internal implementation details should be a last resort.
> 
> I can see a couple of ways this can be done without needing custom code:
> 
> 1. Write the mapping to a local file. The file can be refreshed by specifying `hive.s3.security-mapping.refresh-period`, so you could an external process that periodically pulls the config from your service and writes it to a file.
> 2. Add a `hive.s3.security-mapping.config-uri` that would fetch from an HTTP location rather than a local file. This URL would return the same JSON format. This would allow you to run a simple proxy that fetches the config from your custom service and returns it to Trino.
> 
> Either of these approaches avoids the need to write and maintain custom code that can break between Trino releases. You can integrate with your config service in whatever language and way that is easiest.
> 
> I'm `@david` on the [Trino Slack](https://trino.io/slack.html) if it's easier to discuss there.

In coordination with @david I reworked this PR and added option 2, a `hive.s3.security-mapping.config-uri` fetchinf from an HTTP location. 
I included the option of specifying a `hive.json-pointer` as well to be able to point to the mappings inside the answer from the HTTP location (or the config-file if that is configured). It is completely optional, though, and defaults to just considering the complete JSON response / config-file JSON content.

electrum commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/UriBasedS3SecurityMappingsProvider.java:
```diff
@@ -0,0 +1,68 @@
+        String urlString = config.getConfigUri().orElseThrow(() -> new IllegalArgumentException("hive.s3.security-mapping.config-uri file is not set"));
+        try {
+            URL url = new URL(urlString);
+            HttpURLConnection con = (HttpURLConnection) url.openConnection();
```
We should really use the Airlift HTTP client here, like we do in the rest of Trino, since it integrates with the config system and provides full support for HTTPS configuration rather than relying on JVM defaults. It may feel like overkill here, but provides consistency and makes it more future proof.

You'd do this by changing the bindings in `HiveS3Module` to bind the provider conditionally and inject `S3SecurityMappingsProvider` rather than having `S3SecurityMappingsProviderFactory`. Something like this
```java
private void bindSecurityMapping(Binder binder)
{
    if (buildConfigObject(S3SecurityMappingConfig.class).getConfigFile().isPresent()) {
        binder.bind(S3SecurityMappingsProvider).to(FileBasedS3SecurityMappingsProvider.class).in(Scopes.SINGLETON);
    }
    else if (buildConfigObject(S3SecurityMappingConfig.class).getConfigUri().isPresent()) {
        binder.bind(S3SecurityMappingsProvider).to(UriBasedS3SecurityMappingsProvider.class).in(Scopes.SINGLETON);
        httpClientBinder(binder).bindHttpClient("s3SecurityMapping", ForS3SecurityMapping.class)
                .withConfigDefaults(config -> config
                        .setRequestTimeout(10, SECONDS)
                        .setSelectorCount(1)
                        .setMinThreads(1);
                });
    }
    else {
        return;
    }

    newSetBinder(binder, DynamicConfigurationProvider.class).addBinding()
            .to(S3SecurityMappingConfigurationProvider.class).in(Scopes.SINGLETON);

    checkArgument(!buildConfigObject(HiveConfig.class).isS3SelectPushdownEnabled(), "S3 security mapping is not compatible with S3 Select pushdown");
    checkArgument(!buildConfigObject(RubixEnabledConfig.class).isCacheEnabled(), "S3 security mapping is not compatible with Hive caching");
}
```
See `io.trino.server.security.jwt.JwkService` for a simple example of using the client. You'll need to add an annotation `ForS3SecurityMapping` to qualify the client for injecting (just copy `ForJwk`). Using `createStringResponseHandler()` is probably good for this, just like Jwks.

electrum commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/TestUriBasedS3SecurityMappingsProvider.java:
```diff
@@ -0,0 +1,71 @@
+ */
+package io.trino.plugin.hive.s3;
+
+import com.sun.net.httpserver.HttpServer;
```
This is an internal JDK API that we shouldn't use. See `TestJwkService` for how to use `TestingHttpClient` to test the code after switching to Airlift HTTP client.

electrum commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/S3SecurityMapping.java:
```diff
@@ -96,6 +96,25 @@ public S3SecurityMapping(
         this.endpoint = requireNonNull(endpoint, "endpoint is null");
     }
 
+    private static Predicate<URI> prefixPredicate(URI prefix)
```
Please revert this change, since it's an unrelated refactoring / code organization and not logically part of this PR.

electrum reviewed (commented):
Sorry for the delay reviewing this. This looks good at a high level. I like the use of JSON pointer -- that's a clever way to make the code more flexible. The main comment is about switching to Airlift HTTP client.

clemensvonschwerin commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/S3SecurityMappingsProviderFactory.java:
```diff
@@ -0,0 +1,59 @@
+        List<S3SecurityMappingsProvider> providers = List.of(new FileBasedS3SecurityMappingsProvider(config), new UriBasedS3SecurityMappingsProvider(config));
+        long matching = getNumberOfMatchingProviders(config);
+        if (matching == 0) {
+            throw new IllegalArgumentException("No provider available for current hive.s3.security-mapping configuration. " +
```
Sorry, I do not fully understand your comment. `isProivderAvailable()` should return a boolean indicating if a single, unambiguous provider can be created. Of course, we cannot force devs to call this method first before calling `createMappingsProvider(...)`. Here I wanted to distinguish between ambigous configuration and no provider configured at all as different error causes.

clemensvonschwerin replied:
I guess it does not matter anyways, because I removed `S3SecurityMappingsProviderFactory.java` completely in favour of setup via airlift & binders as you suggested in your comment about using airlift's HttpClient.

clemensvonschwerin commented:
@electrum Thank you for your detailed review, David. I am glad you like the general direction of this PR. I worked through all of your points and updated the PR accordingly. Would be cool, if you could approve running the workflow and hopefully we can merge afterwards.

clemensvonschwerin commented:
Hey @electrum , the checks on the PR passed. Is there anything from your side which remains to be done before merging?

electrum commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/ForS3SecurityMapping.java:
```diff
@@ -0,0 +1,31 @@
+@Retention(RUNTIME)
+@Target({FIELD, PARAMETER, METHOD})
+@BindingAnnotation
+public @interface ForS3SecurityMapping
```
Nit: write empty annotations like
```java
public @interface ForS3SecurityMapping {}

electrum commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/S3SecurityMappingConfig.java:
```diff
@@ -17,13 +17,18 @@
 import io.airlift.configuration.ConfigDescription;
 import io.airlift.configuration.validation.FileExists;
 import io.airlift.units.Duration;
+import org.codehaus.commons.nullanalysis.NotNull;
```
This should be `javax.validation.constraints.NotNull`

electrum commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/S3SecurityMappingConfig.java:
```diff
@@ -42,6 +47,39 @@ public S3SecurityMappingConfig setConfigFile(File configFile)
+    }
+
+    @NotNull
+    public String getJSONPointer()
```
Our code style is to name things in camel case where only the first letter is capitalized, including acronyms, so write this as `getJsonPointer`

electrum commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/S3SecurityMappingConfig.java:
```diff
@@ -42,6 +47,39 @@ public S3SecurityMappingConfig setConfigFile(File configFile)
+
+    @Config("hive.s3.security-mapping.json-pointer")
+    @ConfigDescription("JSON pointer (RFC 6901) to mappings inside JSON config")
+    public S3SecurityMappingConfig setJSONPointer(String jsonPointer)
```
`setJsonPointer`

electrum commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/S3SecurityMappingsProvider.java:
```diff
@@ -0,0 +1,57 @@
+        try {
+            JsonNode node = JsonUtils.parseJson(jsonString.getBytes(StandardCharsets.UTF_8));
+            JsonNode mappingsNode = node.at(this.jsonPointer);
+            return JsonUtils.OBJECT_MAPPER.treeToValue(mappingsNode, S3SecurityMappings.class);
```
Add a utility method `jsonTreeToValue` to `JsonUtils` for this and make `OBJECT_MAPPER` private

electrum commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/S3SecurityMappingsProvider.java:
```diff
@@ -0,0 +1,57 @@
+    protected S3SecurityMappings parse(String jsonString)
+    {
+        try {
+            JsonNode node = JsonUtils.parseJson(jsonString.getBytes(StandardCharsets.UTF_8));
```
Change `parseJson` to take a `String` so that callers don't need to do conversion to bytes. `ObjectMapper` has a `readTree` overload that takes `String` directly.

electrum commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/S3SecurityMappingsProvider.java:
```diff
@@ -0,0 +1,57 @@
+            return JsonUtils.OBJECT_MAPPER.treeToValue(mappingsNode, S3SecurityMappings.class);
+        }
+        catch (JsonProcessingException e) {
+            throw new IllegalArgumentException(format("Failed to parse JSON for S3 security mappings: %s", jsonString), e);
```
```java
throw new UncheckedIOException("Failed to parse JSON for S3 security mappings: " + jsonString, e);

electrum commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/UriBasedS3SecurityMappingsProvider.java:
```diff
@@ -0,0 +1,65 @@
+        extends S3SecurityMappingsProvider
+{
+    private final URI configUri;
+    private HttpClient httpClient;
```
Make final

electrum commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/UriBasedS3SecurityMappingsProvider.java:
```diff
@@ -0,0 +1,65 @@
+    public UriBasedS3SecurityMappingsProvider(S3SecurityMappingConfig config, @ForS3SecurityMapping HttpClient httpClient)
+    {
+        super(config);
+        this.configUri = config.getConfigUri().map(URI::create).orElse(null);
```
This class should only be constructed when the URI is present
```java
this.configUri = config.getConfigUri().orElseThrow(() -> new IllegalArgumentException("configUri not set"));

electrum commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/UriBasedS3SecurityMappingsProvider.java:
```diff
@@ -0,0 +1,65 @@
+    {
+        super(config);
+        this.configUri = config.getConfigUri().map(URI::create).orElse(null);
+        this.httpClient = httpClient;
```
```java
this.httpClient = requireNonNull(httpClient, "httpClient is null");

electrum commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/FileBasedS3SecurityMappingsProvider.java:
```diff
@@ -0,0 +1,56 @@
+    @Override
+    public String getRawJsonString()
+    {
+        if (this.configFile == null) {
```
Not needed after moving validation to constructor

electrum commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/UriBasedS3SecurityMappingsProvider.java:
```diff
@@ -0,0 +1,65 @@
+    @Override
+    protected String getRawJsonString()
+    {
+        if (this.configUri == null) {
```
Not needed after moving validation to constructor

electrum commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/UriBasedS3SecurityMappingsProvider.java:
```diff
@@ -0,0 +1,65 @@
+        if (this.configUri == null) {
+            throw new IllegalArgumentException("hive.s3.security-mapping.config-uri file is not set");
+        }
+        Request request = prepareGet().setUri(this.configUri).build();
```
We normally only use `this.` in the constructor for assignments (often required due to fields and parameters having the same name)

electrum commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/UriBasedS3SecurityMappingsProvider.java:
```diff
@@ -0,0 +1,65 @@
+        try {
+            response = httpClient.execute(request, createStringResponseHandler());
+            int status = response.getStatusCode();
+            if (200 <= status && status <= 299) {
```
I think we only want to accept 200, as anything else won't have a body
```java
if (status != HttpStatusOK.code()) {
    throw new RuntimeException(format("Unexpected status code %s from %s", status, configUri));
}

electrum commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/UriBasedS3SecurityMappingsProvider.java:
```diff
@@ -0,0 +1,65 @@
+            throw new IllegalStateException(format("Request to '%s' returned unexpected status code: '%d'", this.configUri, status));
+        }
+        catch (RuntimeException ex) {
+            throw new IllegalStateException(format("Error while sending get request to '%s'", this.configUri), ex);
```
We shouldn't need to do this, as the HTTP client already throws a good error message on failure

electrum commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/UriBasedS3SecurityMappingsProvider.java:
```diff
@@ -0,0 +1,65 @@
+            throw new IllegalArgumentException("hive.s3.security-mapping.config-uri file is not set");
+        }
+        Request request = prepareGet().setUri(this.configUri).build();
+        StringResponseHandler.StringResponse response;
```
Nit: status import `StringResponse`

electrum commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/TestS3SecurityMappingConfig.java:
```diff
@@ -26,20 +26,30 @@
     }
 
+    @Test
+    public void testDefaultValues()
```
This isn't needed as the `testDefaults` method already does this

electrum commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/TestUriBasedS3SecurityMappingsProvider.java:
```diff
@@ -0,0 +1,43 @@
+    @Test
+    public void testGetRawJSON()
+    {
+        AtomicReference<Response> response = new AtomicReference<>(mockResponse(HttpStatus.OK, JSON_UTF_8, MOCK_MAPPINGS_RESPONSE));
```
I don't think `AtomicReference` is needed here. Doing this as
```java
Response response = ...
new TestingHttpClient(request -> response)
```
should work the same

electrum reviewed (commented):
A few more minor comments. Overall looks great! Apologies for the delay in reviewing

clemensvonschwerin commented:
@electrum thank you for your 2nd review. I addressed the issues you found and would appreciate, if you could approve running the workflow.

clemensvonschwerin commented:
@electrum hey, it has been a while. Is there anything to be done before we can merge this PR?

kokosing commented on docs/src/main/sphinx/connector/hive-s3.rst:
```diff
@@ -189,7 +189,11 @@ In addition to the rules above, the default mapping can contain the optional
-Example JSON configuration file:
+The configuration JSON can either be retrieved from a file specified via
+``hive.s3.security-mapping.config-file`` or from a REST-endpoint specified via
+``hive.s3.security-mapping.config-uri``.
```
Can we make `hive.s3.security-mapping.config-file` to accept URI? Then you don't need to have two properties to maintain. Also the way we read JSON is quite reused (for example for file based access controls) so if your change could be generic enough other places could also benefit from that.

Notice that we already are doing this where one can give an URI for `http-server.authentication.jwt.key-file`.

electrum replied:
My apologies, I've been very busy and haven't followed up here. I appreciate your patience on this. Let me try updating it to use the existing config and get this merged. I do like the idea of a single config.

clemensvonschwerin commented:
Hey @electrum @kokosing, I finally found time to update the PR. Please take another look, I really hope we can get this merged soon.

clemensvonschwerin commented:
@electrum @kokosing it has been a while now, is there a chance that you can take another look at this PR?

kokosing commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/HiveS3Module.java:
```diff
@@ -61,13 +65,29 @@ protected void setup(Binder binder)
+            }
         }
+        else {
+            return;
```
invert the condition, so `else` block is not needed

kokosing commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/S3SecurityMappingConfigurationProvider.java:
```diff
@@ -57,35 +55,33 @@
         return Suppliers.memoizeWithExpiration(
                 () -> {
-                    log.info("Refreshing S3 security mapping configuration from %s", configFile);
+                    log.info("Refreshing S3 security mapping configuration");
```
Why removed?

kokosing commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/S3SecurityMappingsProvider.java:
```diff
@@ -0,0 +1,47 @@
+
+import static java.util.Objects.requireNonNull;
+
+public abstract class S3SecurityMappingsProvider
```
Can we use composition instead of inheritance here?

clemensvonschwerin replied:
How exactly would you like that to look? Remove `abstract` from `S3SecurityMappingsProvider` and create a new interface for `UriBasedS3SecurityMappingsProvider` and `FileBasedS3SecurityMappingsProvider` to implement?

kokosing replied:
I think it is fine to have an interface for `S3SecurityMappingsProvider` that is only `implements Supplier<S3SecurityMappings>`, but then when implementing the other classes we could extract the common code to a separate class that is used from `UriBasedS3SecurityMappingsProvider` and `FileBasedS3SecurityMappingsProvider`

clemensvonschwerin replied:
Alright, I will update the PR accordingly

kokosing reviewed (commented):
Left few comments. I am very sorry for a delay. 

Please try to follow advices from: https://github.com/trinodb/trino/pull/8411

clemensvonschwerin commented:
@kokosing @electrum I updated the PR and tried multiple reruns, the oauth2 checks do not seem to be happy for some reason I know nothing about. What can I do?

cvs-mckinsey commented:
@kokosing @electrum Hey guys, I think we are super close to merging this but there has been no activity for some weeks now. Could you please take a look?

kokosing commented on docs/src/main/sphinx/connector/hive-s3.rst:
```diff
@@ -169,7 +169,7 @@ The security mapping must provide one or more configuration settings:
   credential. If list cotains "*", then any key can be specified via extra credential.
```
Please structure commits according to: https://github.com/trinodb/trino/blob/master/DEVELOPMENT.md#git-merge-strategy

Git history matters for us a lot, and this is also a matter of code review.

cvs-mckinsey replied:
Sure, after all your comments are resolved I will squash all my commits into a single one and commit that one

kokosing commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/FileBasedS3SecurityMappingsProvider.java:
```diff
@@ -0,0 +1,57 @@
+    {
+        this.configFile = config.getConfigFilePath().map(File::new).orElseThrow(() -> new IllegalArgumentException("configFile not set"));
+        if (!configFile.exists()) {
+            throw new IllegalArgumentException(String.format("configFile %s does not exist", configFile.getAbsoluteFile()));
```
static import for `format`

cvs-mckinsey replied:
Updated

kokosing commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/FileBasedS3SecurityMappingsProvider.java:
```diff
@@ -0,0 +1,57 @@
+    @Override
+    public S3SecurityMappings get()
+    {
+        return parser.parseJSONString(getRawJsonString());
```
are we going to download and load mapping every single call? It might contribute the query latency. I think we need to have some kind of caching with usage of `Suppliers.memoizeWithExpire` for this.

cvs-mckinsey replied:
Isn't that already handled by `S3SecurityMappingConfigurationProvider::getMappings(...)` ?

kokosing commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/HiveS3Module.java:
```diff
@@ -61,13 +65,28 @@ protected void setup(Binder binder)
+                    .withConfigDefaults(config -> config
+                            .setRequestTimeout(Duration.succinctDuration(10, TimeUnit.SECONDS))
+                            .setSelectorCount(1)
+                            .setMinThreads(1));
```
Why these are set to `1`?

cvs-mckinsey replied:
Taken from @electrum 's suggestion in https://github.com/trinodb/trino/pull/6210#discussion_r620687229 
If you have any other suggestions I can update these values accordingly or just use the defaults.

kokosing commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/TestUriBasedS3SecurityMappingsProvider.java:
```diff
@@ -0,0 +1,41 @@
+    {
+        final Response response = mockResponse(HttpStatus.OK, JSON_UTF_8, MOCK_MAPPINGS_RESPONSE);
+        S3SecurityMappingConfig conf = new S3SecurityMappingConfig().setConfigFilePath("http://test:1234/api/endpoint");
+        UriBasedS3SecurityMappingsProvider provider =
```
No need to wrap lines here.

cvs-mckinsey replied:
Updated

kokosing reviewed (commented):
LGTM % comments.

kokosing commented on lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/JsonUtils.java:
```diff
@@ -50,13 +57,30 @@ private JsonUtils() {}
         }
     }
 
+    public static JsonNode parseJson(String jsonStr)
```
Please rename `jsonStr` to just `json`

cvs-mckinsey commented:
> LGTM % comments.

@kokosing @electrum I updated the PR based on the review comments and squashed all commits into a single one for a clean git history

kokosing commented on docs/src/main/sphinx/connector/hive-s3.rst:
```diff
@@ -169,7 +169,7 @@ The security mapping must provide one or more configuration settings:
   credential. If list cotains "*", then any key can be specified via extra credential.
```
Can you please format your commit message according to: https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#format-git-commit-messages?

cvs-mckinsey replied:
Done

kokosing reviewed (approved):
% comment

kokosing commented:
Merged, thank you!

clemensvonschwerin commented:
> Merged, thank you!

Awesome, thank you!
